sound/soc/sof/topology.c

```diff
@@ -1515,6 +1515,11 @@ static int sof_widget_ready(struct snd_soc_component *scomp, int index,
 		break;
 	}
 
+	if (swidget->id == snd_soc_dapm_aif_in || swidget->id == snd_soc_dapm_dai_in)
+		swidget->dir = SNDRV_PCM_STREAM_PLAYBACK;
+	else if (swidget->id == snd_soc_dapm_aif_out || swidget->id == snd_soc_dapm_dai_out)
+		swidget->dir = SNDRV_PCM_STREAM_CAPTURE;
```

Comment:
This seems completely unnecessary. If you are setting the dir based on id, you've already got all the information you need. Why do you need this at all?

Reply:
> This seems completely unnecessary. If you are setting the dir based on id, you've already got all the information you need. Why do you need this at all?

We can only know the direction if the widget type is dai or aif. For example, we don't know the direction if the widget type is snd_soc_dapm_pga. This is the first step to assign swidget direction, the second step is to assign the same direction to their pipe_widget, and assign the same direction to all swidgets associated to the pipe_widget.